sqlglot/parser.py

```diff
@@ -5571,12 +5569,15 @@ def _parse_not_constraint(self) -> t.Optional[exp.Expression]:
         return None
 
     def _parse_column_constraint(self) -> t.Optional[exp.Expression]:
-        if self._match(TokenType.CONSTRAINT):
-            this = self._parse_id_var()
-        else:
-            this = None
+        this = self._match(TokenType.CONSTRAINT) and self._parse_id_var()
+
+        index = self._index
+        procedure_property_follows = self._match(TokenType.WITH) and self._match_texts(
+            self.PROCEDURE_OPTIONS
+        )
+        self._retreat(index)
```

Comment:
Let's retreat conditionally, only if the flag is true

Reply:
I need to keep the index where it is because I need `WITH` to be the current token when I get to the property parsers where I parse the procedure options. That I why I was always retreating. 

I could rewrite the condition as:
```
procedure_property_follows = self._match(TokenType.WITH, advance=False) and self._next.text.upper() in self.PROCEDURE_OPTIONS
```
That way I avoid having to advance on matches and retreat. I didn't do it like that at first because I preferred using `self._match_texts(self.PROCEDURE_OPTIONS)` instead of `self._next.text.upper() in self.PROCEDURE_OPTIONS`.